k2/csrc/cuda/debug.h

```diff
@@ -0,0 +1,383 @@
+ *  `_K2CudaDebug`.
+ *
+ * @todo
+ *  Should the error code return by this macro?
+ *
+ * @code{.cpp}
+ * K2_CUDA_CHECK_ERROR(cudaGetLastError());
+ * @endcode
+ */
+#ifndef K2_CUDA_CHECK_ERROR
+  #ifdef K2_MAKE_ERROR_CHECK
+    #define K2_CUDA_CHECK_ERROR(e)                                    \
+      if (::k2::_K2CudaDebug((cudaError_t)(e), __FILE__, __LINE__)) { \
+        exit(e);                                                      \
+      }
+  #else
+    #define K2_CUDA_CHECK_ERROR(e) ((void)0)
+  #endif
+#endif
+
+/**
+ * @def K2_CUDA_API_SAFE_CALL(CUDA_RUNTIME_API)
+ *
+ * @brief Macro for checking cuda standard runtime api return status.
+ *
+ * @details
+ *  If api return status is not cudaSuccess, print and exit.
+ *  Thus it's actually could be an alias of K2_CUDA_CHECK_ERROR.
+ *
+ * @note
+ *  All runtime api return an error code. But for an asynchronous api,
+ *  the error code only reports errors that occur on the host
+ *  prior to executing the task, typically related to parameter validation.
+ *  The device kernel error would unfortunately be left for next runtime api
+ *  call. To check the whole process of cuda runtime async api,
+ *  as cuda kernel launches are asynchronous also,
+ *  the same macro for kernel could be used. See
+ *  `K2_CUDA_KERNEL_SAFE_CALL`.
+ *
+ * @code{.cpp}
+ * K2_CUDA_API_SAFE_CALL(CUDA_RUNTIME_API());
+ * @endcode
+ */
+#ifndef K2_CUDA_API_SAFE_CALL
+  #ifdef K2_MAKE_ERROR_CHECK
+    #define K2_CUDA_API_SAFE_CALL(c) \
+      do {                           \
+        cudaError_t e = (c);         \
+        K2_CUDA_CHECK_ERROR(e);      \
+      } while (0)
+  #else
+    #define K2_CUDA_API_SAFE_CALL(c) (c)
+  #endif
+#endif
+
+/**
+ * @def K2_CUDA_KERNEL_SAFE_CALL([kernel|cuda_runtime_api])
+ *
+ * @brief Macro for checking cuda standard runtime api return status.
+ *
+ * @details
+ *  If api return status is not cudaSuccess, print and exit.
+ *  Thus it's actually an alias of K2_CUDA_CHECK_ERROR.
+ *
+ * @note
+ *  Kernel launches do not return any error code, thus checking should after it.
+ *  To wait kernel to finish, cudaDeviceSynchronize is called between.
+ *  And to avoid pre-launch error disturbing, check error before kernel launch
+ *  is also needed.
+ *  (But not necessary if programmer promise each error is take cared properly.)
+ *
+ * @remark
+ *  macro `__VA_ARGS__` is used to pass the kernel<<<...>>> as one argument,
+ *  otherwise the compiler raises a error "passed 2 arguments, but takes just 1".
+ *
+ * @code{.cpp}
+ * K2_CUDA_KERNEL_SAFE_CALL(kernel_func<<<...>>>());
+ * @endcode
+ */
+#ifndef K2_CUDA_KERNEL_SAFE_CALL
+  #ifdef K2_MAKE_ERROR_CHECK
+    #define K2_CUDA_KERNEL_SAFE_CALL(...)               \
+      do {                                              \
+        K2_CUDA_CHECK_ERROR(cudaGetLastError());        \
+        (__VA_ARGS__);                                  \
+        K2_CUDA_API_SAFE_CALL(cudaDeviceSynchronize()); \
```

Comment:
Even if NDEBUG is defined I think we should still be checking for the cuda error code
(I think it's OK to do this either before or after the kernel invocation, but not both).

Reply:
After `GetLastError()`, the error flag is set to cudaSuccess. Thus, check before a kernel is to rule out before errors. As the kernel is async, it should not be in production codes. As to what you mean, the `K2_MAKE_ERROR_CHECK` is used in this situation, define which manually in any situation. These error checks are turned on. And by default, its effect is the same as `NDEBUG`, I mean it only activate in DEBUG mode.

Reply:
OK, can you please simplify this by removing K2_MAKE_ERROR_CHECK and just relying on NDEBUG?  Less confusing.
But I want you to check the cuda error codes from kernels once (say, after the kernel invocation, although it doesn't matter) even if NDEBUG is defined, because we need to catch errors at *some* point, even if late.

Reply:
Maybe I didn't express clearly, but what I meant was that `K2_MAKE_ERROR_CHECK` is just what satisfies what you're saying. `K2_MAKE_ERROR_CHECK` regardless of whether `NDEBUG` is defined or not, if `K2_MAKE_ERROR_CHECK` is defined manually, then it will definitely do error checking.

Reply:
Check only once after kernel is ok. Changes are made.